### PR TITLE
Implement the new-style trait solver

### PR DESCRIPTION
**Pull request by scalexm**
Final PR of what I believe to be a minimally working implementation of the new-style trait solver.

The new trait solver can be used by providing the `-Z chalk` command line flag. It is currently used everywhere in `rustc_typeck`, and for everything relying on `rustc::infer::canonical::query_response::enter_canonical_trait_query`.

The trait solver is invoked in rustc by using the `evaluate_goal` canonical query. This is not optimal because each call to `evaluate_goal` creates a new `chalk_engine::Forest`, hence rustc cannot use answers to intermediate goals produced by the root goal. We'll need to change that but I guess that's ok for now.

Some next steps, I think, are:
* handle region constraints: region constraints are computed but are completely ignored for now, I think we may need additional support from `chalk_engine` (as a side effect, types or trait references with outlive requirements cannot be proved well-formed)
* deactivate eager normalization in the presence of `-Z chalk` in order to leverage the lazy normalization strategy of the new-style trait solver
* add the remaining built-in impls (only `Sized` is supported currently)
* transition the compiler to using generic goals instead of predicates that still refer to named type parameters etc

I added a few very simple tests to check that the new solver has the right behavior, they won't be needed anymore once it is mature enough. Additionally it shows off that we get [implied bounds](https://github.com/rust-lang/rust/issues/44491) for free.

r? @nikomatsakis

### CODE REVIEW

**Comment by mark-i-m** [comment]
The title of this PR is so epic

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Sorry for being so slow @scalexm -- this looks great, but I do have a few concerns:

- We should decide how to handle subtyping before landing; a FIXME may suffice here
- I'm a *bit* concerned about adding the def-id to param-env, it seems like this could affect queries in unintended ways, leading to less re-use than we would otherwise get

(Maybe we can only set the def-id if `-Zchalk` is used for now?)

File: src/librustc_traits/chalk_context/mod.rs
Comment (scalexm) [reply src/librustc_traits/chalk_context/mod.rs]: https://github.com/rust-lang-nursery/chalk/pull/191

File: src/librustc_traits/chalk_context/mod.rs
Comment (scalexm) [reply src/librustc_traits/chalk_context/mod.rs]: Yes it’s a hack, I was planning to remove it from the final commit history.

File: src/librustc_traits/chalk_context/mod.rs
Comment (scalexm) [reply src/librustc_traits/chalk_context/mod.rs]: There are some lost in the `chalk-integration` Zulip channel :)

I’ll elaborate on that soon.

File: src/librustc_traits/chalk_context/program_clauses.rs
Comment (scalexm) [reply src/librustc_traits/chalk_context/program_clauses.rs]: This is definitely a hack indeed. We need to add a FIXME yes.

Comment (scalexm) [reply src/librustc_traits/chalk_context/program_clauses.rs]: Also I believe this problem is indeed not unique to `Sized`, but at least it would only show for traits that have built-in impls generated by the compiler.

**Comment by nikomatsakis** [comment]
Current status to my knowledge:

Waiting on @scalexm to rebase over the new version of chalk, at least, and perhaps to answer a few more of the questions above.

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@bors r+

I think it's a good idea to land this so we can iterate in tree

**Pull request merged**
